10_predictive-modeling-review.qmd

```diff
@@ -0,0 +1,1045 @@
+
+New means and standard deviations need to be estimated every time a model is fit including within each iteration of a resampling method! This is a lot of work, but `library(tidymodels)` makes this simple.
+:::
+
+## Predictive Modeling Pipelines
+
+The rest of this chapter focuses on a predictive modeling pipeline applied to the simulated data. This pipeline is a general approach to creating predictive models and can change based on the specifics of an application. We will demonstrate the pipeline using KNN and regression trees. 
+
+
+0. Problem Formulation
+1. Split data into training and testing data
+2. Exploratory Data Analysis
+3. Set up Resampling for Model Selection
+4. Create Candidate Models
+5. Test and Choose the "Best" Model
+6. Optional: Expand the Model Fit
+7. Evaluate the Final Model
+8. Optional: Implement the Final Model
+
+
+### Example: KNN
+
+#### 1. Split data into training and testing data
+
+```{r}
+set.seed(20201007)
+
+# create a split object
+data1_split <- initial_split(data = data1, prop = 0.75)
+
+# create the training and testing data
+data1_train <- training(x = data1_split)
+data1_test  <- testing(x = data1_split)
+
+```
+
+#### 2. Exploratory Data Analysis
+
+:::{.callout-warning}
+Only perform EDA on the training data. 
+
+Exploring the testing data will lead to data leakage. 
+:::
+
+```{r}
+data1_train |>
+  ggplot(aes(x = x, y = y)) +
+  geom_point(alpha = 0.25) +
+  labs(title = "Example 1 Data") +
+  theme_minimal()
+
+```
+
+#### 3. Set up Resampling for Model Selection
+
+We use 10-fold cross validation for hyperparameter tuning and model selection. The training data are small (750 observations and one predictor), so we will repeat the entire cross validation process five times.
+
+Estimating out-of-sample error rates with cross-validation is an uncertain process. Repeated cross-validation improves the accuracy of our estimated error rates.
+
+```{r}
+data1_folds <- vfold_cv(data = data1_train, v = 10, repeats = 5)
+
+```
+
+#### 4. Create Candidate Models
+
+We have three main levers for creating candidate models:
+
+1. **Feature and target engineering.** Feature and target engineering is generally specified when creating a recipe with `recipe()` and `step_*()` functions. 
+2. **Switching algorithms.** Algorithms are specified using functions from `library(parsnip)`.
+3. **Hyperparameter tuning.** Hyperparameter tuning is typically handled by using `tune()` when picking an algorithm and then creating a hyperparameter tuning grid. 
+
+It is common to normalize (subtract the sample mean and divide by the sample standard deviation) predictors when using KNN. 
+
+```{r}
+knn_recipe <-
+  recipe(formula = y ~ ., data = data1_train) |>
+  step_normalize(all_predictors())
+
+```
+
+We use `nearest_neighbor()` to create a KNN model and we use `tune()` as a placeholder for `k`.
```

Comment:
I could use a little more clarity here. I would say something like:

Note that you can tune many hyperparameters for a given model. The Regression Tree example later in this chapter illustrates this concept.

Reply:
Added.